apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerConnection.cs

```diff
@@ -0,0 +1,313 @@
+        {
+        }
+
+        /// <summary>
+        /// Creates a SpannerConnection with a datasource contained in connectionString
+        /// and optional credential information supplied in connectionString or the credential
+        /// argument.
+        /// </summary>
+        /// <param name="connectionString">A Spanner formatted connection string.</param>
+        /// <param name="credential">An optional credential.</param>
+        public SpannerConnection(string connectionString, ITokenAccess credential = null)
+            :this(new SpannerConnectionStringBuilder(connectionString, credential))
+        {
+        }
+
+        /// <summary>
+        /// Creates a SpannerConnection with a datasource contained in connectionString
+        /// and optional credential information supplied in connectionString or the credential
+        /// argument.
+        /// </summary>
+        /// <param name="connectionStringBuilder">A SpannerConnectionStringBuilder containing
+        ///  a formatted connection string.</param>
+        public SpannerConnection(SpannerConnectionStringBuilder connectionStringBuilder)
+        {
+            TrySetNewConnectionInfo(connectionStringBuilder);
+        }
+
+        private bool IsOpen => (State & ConnectionState.Open) == ConnectionState.Open;
+        private bool IsClosed => (State & ConnectionState.Open) == 0;
+
+        private void AssertOpen(string message) {
+            if (!IsOpen)
+            {
+                throw new InvalidOperationException("The connection must be open.  Failed to " + message);
+            }
+
+        }
+
+        private void AssertClosed(string message) {
+            if (!IsClosed)
+            {
+                throw new InvalidOperationException("The connection must be closed.  Failed to " + message);
+            }
+
+        }
+        private void TrySetNewConnectionInfo(SpannerConnectionStringBuilder newBuilder) {
+            AssertClosed("change connection information.");
+            _connectionStringBuilder = newBuilder;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public SpannerBatchOperation BeginBatchOperation()
+        {
+            //This may be called while the connection is closed.
+            SpannerBatchOperation newBatchOperation = new SpannerBatchOperation(this);
+            if (Interlocked.CompareExchange(ref _activeBatchOperation, newBatchOperation, null)
+                != null) {
+                throw new InvalidOperationException("You may only create a single Batch operation at a time.");
+            }
+            return _activeBatchOperation;
+        }
+
+#if NET451
+        /// <inheritdoc />
+        public override void EnlistTransaction(System.Transactions.Transaction transaction)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc />
+        protected override DbProviderFactory DbProviderFactory
```

Comment:
actually this PR is a bit old as I've reformatted those that can be expression-bodied.  will update today